src/goto-programs/slice_global_inits.cpp

```diff
@@ -0,0 +1,143 @@
+
+ Purpose:
+
+\*******************************************************************/
+
+void slice_global_inits(
+  const namespacet &ns,
+  goto_functionst &goto_functions)
+{
+  // gather all functions reachable from the entry point
+
+  call_grapht call_graph(goto_functions);
+  const call_grapht::grapht &graph=call_graph.graph;
+
+  std::list<irep_idt> worklist;
+  std::set<irep_idt> functions_reached;
+
+  const irep_idt entry_point=goto_functionst::entry_point();
+
+  goto_functionst::function_mapt::const_iterator e_it;
+  e_it=goto_functions.function_map.find(entry_point);
+
+  if(e_it==goto_functions.function_map.end())
+    throw "entry point not found";
+
+  worklist.push_back(entry_point);
+
+  do
+  {
+    const irep_idt id=worklist.front();
+    worklist.pop_front();
+
+    functions_reached.insert(id);
+
+    const auto &p=graph.equal_range(id);
+
+    for(auto it=p.first; it!=p.second; it++)
+    {
+      const irep_idt callee=it->second;
+
+      if(functions_reached.find(callee)==functions_reached.end())
+        worklist.push_back(callee);
+    }
+  } while(!worklist.empty());
+
+  const irep_idt initialize=CPROVER_PREFIX "initialize";
+  functions_reached.erase(initialize);
+
+  // gather all symbols used by reachable functions
+
+  class symbol_collectort:public const_expr_visitort
+  {
+  public:
+    virtual void operator()(const exprt &expr)
+    {
+      if(expr.id()==ID_symbol)
+      {
+        const symbol_exprt &symbol_expr=to_symbol_expr(expr);
+        const irep_idt id=symbol_expr.get_identifier();
+        symbols.insert(id);
+      }
+    }
+
+    std::set<irep_idt> symbols;
+  };
+
+  symbol_collectort visitor;
+
+  assert(!functions_reached.empty());
+
+  for(std::set<irep_idt>::const_iterator id_it=functions_reached.begin();
+      id_it!=functions_reached.end(); id_it++)
+  {
+    const irep_idt id=*id_it;
+    const goto_functionst::goto_functiont &goto_function
+      =goto_functions.function_map.at(id);
+    const goto_programt &goto_program=goto_function.body;
+
+    forall_goto_program_instructions(i_it, goto_program)
+    {
+      const codet &code=i_it->code;
+      code.visit(visitor);
+    }
+  }
+
+  const std::set<irep_idt> &symbols=visitor.symbols;
+
+  // now remove unnecessary initializations
+
+  goto_functionst::function_mapt::iterator f_it;
+  f_it=goto_functions.function_map.find(initialize);
+  assert(f_it!=goto_functions.function_map.end());
+
+  goto_programt &goto_program=f_it->second.body;
+
+  Forall_goto_program_instructions(i_it, goto_program)
+  {
+    if(i_it->is_assign())
+    {
+      const code_assignt &code_assign=to_code_assign(i_it->code);
+      const symbol_exprt &symbol_expr=to_symbol_expr(code_assign.lhs());
+      const irep_idt id=symbol_expr.get_identifier();
+
+      if(has_prefix(id2string(id), CPROVER_PREFIX))
+        continue;
+
+      if(symbols.find(id)!=symbols.end())
+        continue;
+
+      i_it->make_skip();
+    }
+  }
+
+  goto_functions.update();
```

Comment:
Do the remove_skip before the update right here, instead of at the point of call of this overall code.